### ISSUE TITLE
is it possible to load tiles in parallel while using sequence mode?

### ISSUE THREAD

**Issue by atacraft** [openseadragon/openseadragon#761]
Hi!

Just asking ... is it possible to have parallel download when using sequence mode (as it is said in the documentation "Set to true to have the viewer treat your tilesources as a sequence of images to be opened one at a time rather than all at once.").

here are screenshots of the network monitor with my demo ...

### sequenceMode: true
![loading_sequence_true](https://cloud.githubusercontent.com/assets/2459696/10908308/e7194c9e-8228-11e5-9d3d-df03e6008022.png)

### sequenceMode: false

![loading_sequence_false](https://cloud.githubusercontent.com/assets/2459696/10908318/f71d9f5a-8228-11e5-8a9e-a8c8bde8d65a.png)

**Comment by avandecreme** [comment]
No, OpenSeadragon seems to be starting one download per tile source in both modes.
The only difference is that with sequenceMode to false, you have a lot of tile sources which are loading in parallel. With sequenceMode to true, you only have one tile source open at a time.
I am not sure why OSD start the downloads sequentially like this though. Maybe something which can be improved.

**Comment by atacraft** [comment]
As you mentionned OSD launches as many tile's download as he can in parallel when sequenceMode is false. My concern is about having this feature while using sequenceMode (true).

It's just my opinion but i think we shouldn't be waiting for a tile download to finish before asking for another one ... and when we are in sequenceMode, when someone is viewing a given page he is likely to click on next, previous or zoomIn button ... why do we have to wait until he ask before we launch the download of needed tiles ...

After all answers to those question may lie under business requirements but I think one should be able to choose.

**Comment by iangilman** [comment]
Looks like there are two questions here: 

* Why aren't the tiles for a single tiled image loading in parallel during sequenceMode? 
* Why don't we load tiles for images that aren't currently displayed during sequenceMode? 

For the first question, I'm not sure why it's not doing it for you...they load in parallel for me, and that's what they're programmed to do. Unless you've set the `imageLoaderLimit` option to 1, which limits them to serial. Can you share the options you're using, or share a link to your app? 

For the second question, OpenSeadragon restricts itself to only loading tiles that are visible. The library is designed to deal with literally gigabytes worth of data, sometimes over limited data connections like mobile phones, so sticking to only the visible tiles is a good policy. In sequenceMode, the next image is not yet visible, so it doesn't load. 

That said, being able to selectively preload not-yet-visible tiles is a popular feature request. @atacraft  would you be interested in exploring a patch for that? I think it should be an optional feature to opt into. 

Incidentally, we're contemplating moving sequenceMode out into its own plugin (see #649) where it could possibly evolve more rapidly.

**Comment by atacraft** [comment]
Yes, two questions.

* No link available yet ... but here comes my options:

```
var zoomViewerOptions = {
		id: "zoomOpenseadragon",
		prefixUrl: "/openseadragon/images/",
		tileSources: tiles_tab,
		sequenceMode: true,
		navPrevNextWrap: true,
		immediateRender: false,
		useCanvas: false,
		maxZoomLevel: 1,
		zoomPerClick: 2,
		zoomInButton:   "zoom-in",
		zoomOutButton:  "zoom-out",
		homeButton:     "home",
		fullPageButton: "full-page",
		nextButton:     "next",
		previousButton: "previous",
		gestureSettingsMouse: {
			scrollToZoom: true,
			clickToZoom: false,
			dblClickToZoom: true
		},
		gestureSettingsTouch: {
			scrollToZoom: false,
			clickToZoom: false,
			dblClickToZoom: true
		},
		gestureSettingsPen: {
			scrollToZoom: false,
			clickToZoom: false,
			dblClickToZoom: true
		},
		gestureSettingsUnknown: {
			scrollToZoom: false,
			clickToZoom: false,
			dblClickToZoom: true
		}
	};
```

As you see ```imageLoaderLimit``` is not set thus it should be zero (the default value).

Investigating into viewer.js, open function (starting at line 478) i found this block of code:

```
if (this.sequenceMode && $.isArray(tileSources)) {
            if (this.referenceStrip) {
                this.referenceStrip.destroy();
                this.referenceStrip = null;
            }

            this.tileSources = tileSources;
            this._sequenceIndex = Math.max(0, Math.min(this.tileSources.length - 1, this.initialPage));
            if (this.tileSources.length) {
                this.open(this.tileSources[this._sequenceIndex]);

                if ( this.showReferenceStrip ){
                    this.referenceStrip = new $.ReferenceStrip({
                        id:          this.referenceStripElement,
                        position:    this.referenceStripPosition,
                        sizeRatio:   this.referenceStripSizeRatio,
                        scroll:      this.referenceStripScroll,
                        height:      this.referenceStripHeight,
                        width:       this.referenceStripWidth,
                        tileSources: this.tileSources,
                        prefixUrl:   this.prefixUrl,
                        viewer:      this
                    });
                }
            }

            this._updateSequenceButtons( this._sequenceIndex );
            return;
        }
```

as you can see with ```this.open(this.tileSources[this._sequenceIndex]);``` only one image will be opened ... it's not the case when sequenceMode is false ... inside thee same open function we reach this block of code:

```
for (var i = 0; i < tileSources.length; i++) {
            doOne(tileSources[i]);
        }
```

Iterating over all tiled images and starting to download associated tiles for them, launching as many requests as possible.

Maybe i'm wrong ... but here is a hint i got while hacking around ... adding pieces of code inside the ```if (this.tileSources.length) {...}``` of ```if (this.sequenceMode && $.isArray(tileSources)) {...}``` block resulted in parallel download and even loading tiles that are not displayed yet (it was just hacking to see what's happening :-) ).

* It's not a promise (as i'm working on two projects and moving from one to another depending on what's needed) ... but i'll take a look as soon as possible (if someone is in hurry, he'd better dive into it ^_^)

**Comment by iangilman** [comment]
Hmm, I don't see anything in those options that would explain why it's loading your single image tiles in sequence. 

Anyway, you seem more concerned about the fact that it's not loading tiles from other images in the sequence. As I mentioned, this is an intentional design choice. On the other hand, if/when we break sequenceMode out into its own plugin, it would be great to add additional options to it. 

By adding all the tiled images to the viewer at the same time (the effect you get by turning sequenceMode off), not only does it give you more control over which is loaded when, but it also opens up the possibility of transitions between images (such as fade or slide). All of that would be great, but it's enough of a re-architecting that I'd want us to do it in the new plugin rather than here in OSD core. 

If you're interested in diving in and getting this new sequenceMode plugin rolling, that would be awesome!

**Comment by atacraft** [comment]
Yes, i am ... it would be great to do so, i'll let you know as soon as possible.

**Comment by iangilman** [comment]
Awesome :)
